### ISSUE TITLE
Continuously created new accounts with OAuth

### ISSUE THREAD

**Issue by powderluv** [davido/gerrit-oauth-provider#69]
I am running the emergency fix of Gerrit 2.13.1 in the release page and the OAuth plugin. Now everytime someone who has never logged in before logs in a new account is created. So if they sign out and sign in a new account is created. I used the gsql command interface to see a stream of new accounts created.

 Old accounts (upgraded from 2.12.x) work fine.

**Comment by powderluv** [comment]
BTW this doesn't happen when creating a new account using Github OAuth - only Google OAuth suffers from the problem.

**Comment by tprochazka** [comment]
We have the same problem. Ans also it is not possible to create username, it is blank a there is no way to set it up. I think that it cause that new account is created every time,because old one is not valid because of missing username. We also have emergency fix and using Gooogle OAuth.

**Comment by westhusing** [comment]
I am running the emergency fix build of Gerrit (2.13.1-10-gbf8b34a-dirty) and I have not seen this.  What table are you looking at to see the new accounts?

**Comment by powderluv** [comment]
@tprochazka I noticed the inability to create the username too. I tried adding it via gsql into the account_external_ids and it didn't help.

Here was my command:
gerrit> insert into account_external_ids (account_id, external_id) values (1000178, 'username:FFF');
UPDATE 1; 0 ms

@westhusing I checked the accounts table

Here is a sample output (I hand edited the domain name and replaced with FFF to prevent spam):

 powder luv            | N        | powderluv@powderFFF.org  | 2016-10-03 21:58:17.382 | 1000178
 powder luv            | N        | powderluv@powderFFF.org  | 2016-10-03 21:58:31.683 | 1000179

**Comment by westhusing** [comment]
@powderluv That's the same gsql command I ran to update the username.  I also had to restart Gerrit to get the database changes to be recognized.

Yeah, the instance of Gerrit that I am running is not creating multiple accounts.  I am running with Google OAuth as well.

**Comment by tprochazka** [comment]
We solved it by update this oauth plugin configuration, it was outdated
But I don't know exact details didn't do it.

**Comment by davido** [comment]
I don't have time to investigate the problem atm, but I think that two issues are unrelated:

* multiple users create for the same account (that sounds very strange to me, as the key is external id Google OAuth number and nothing more)
* not able to set usernamem any more for new accounts

I know that I made it work in this change: [1]. This was apparently broken in the same series, that caused the NPE, that was recently fixed: [2].

Later it was extended to support editing (again), but only a subset of fields: [3].

In short it's broken now and usernames can only be set through database manipulation for new accounts until it's fixed.

[1] https://gerrit-review.googlesource.com/66311
[2] https://gerrit-review.googlesource.com/71735
[3] https://gerrit-review.googlesource.com/74441

**Comment by Nathan-Wall** [comment]
I am using https://github.com/davido/gerrit-oauth-provider/releases/download/v2.13.1/gerrit-v2.13.1-10-gbf8b34a.war with https://github.com/davido/gerrit-oauth-provider/releases/download/v2.13.1/gerrit-oauth-provider.jar and am having this problem.

I don't have any idea if this is related, but I also have a user created with [github-plugin](https://gerrit.googlesource.com/plugins/github/+/master/README.md) -- that may be a little unusual.

**Comment by Nathan-Wall** [comment]
> that sounds very strange to me, as the key is external id Google OAuth number and nothing more

When I sign out and back in using Google OAuth, the old record from `acount_external_ids` drops and a new one is inserted (or at least the `account_id` field is updated to a new value).

**Comment by davido** [comment]
So, I can reproduce this problem now. It only happens for new accounts, that were created after migration from 2.12 to 2.13. Looks like the reason for it is new secondary index for accounts, that was introduced in 2.13. When new user is created, at least in OAuth code path, the reindexing doesn't happens, and after logout/login sequence, the existing user is not found in the secondary index, and thus re-created. So this seems to be a regression of newly introduced account index in Gerrit core and not related to this plugin.

As always in this case, this should be tracked in upstream issue tracker.

**Comment by powderluv** [comment]
@davido thanks for the explanation. Would you be able to file the issue upstream or would you like me to do it. 

Also why doesn't this affect Github OAuth ? That is why I was suspecting the plugin.

**Comment by davido** [comment]
I will file issue upstream in a moment, still investigating.

> Also why doesn't this affect Github OAuth ?

I think it should affect GH OAuth as well, if user is created after the migration and thus not contained in the account index yet.

**Comment by davido** [comment]
Fixed in: [1].

*  https://gerrit-review.googlesource.com/89270

**Comment by davido** [comment]
Given that this is fixed now, i filed another issue upstream for broken user name mutation for OAuth scheme: [1].

* [1] https://bugs.chromium.org/p/gerrit/issues/detail?id=4784

**Comment by powderluv** [comment]
@davido Thanks for the quick fix. Does https://gerrit-review.googlesource.com/89270 needs some reviewers to be added to get merged ?

**Comment by davido** [comment]
The fix for hopefully last OAuth extension point regression in 2.13 is here: [1].

* [1] https://gerrit-review.googlesource.com/89291

**Comment by davido** [comment]
@powderluv 

> Does https://gerrit-review.googlesource.com/89270 needs some reviewers to be added to get merged ?

It was already merged :smile:.

**Comment by davido** [comment]
Now, that all fixes were merged upstream, we are trying to release 2.13.2 ASAP: [1].

* [1] https://gerrit-review.googlesource.com/89311

**Comment by davido** [comment]
I've updated the emergency fix and uploaded it in this release: [1].

* [1] https://github.com/davido/gerrit-oauth-provider/releases/tag/v2.13.1
